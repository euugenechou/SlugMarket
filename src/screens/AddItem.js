/* React imports */
import React from "react";
import Expo from "expo";
import {
  TextInput,
  StyleSheet,
  View,
  Button,
  ScrollView,
  Alert,
  Picker
} from "react-native";
import { API, Auth, Storage } from "aws-amplify";

export default class AddItem extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      userId: "",
      category: "",
      description: "",
      isSold: false,
      itemName: "",
      price: 0,
      seller: "",
      timeAdded: ""
    };
  }

  onChangeText(key, value) {
    this.setState({
      [key]: value
    });
  }

  /**
   * Open image picker for creating a new post
   */
  async getPostImage() {
    let result = await Expo.ImagePicker.launchCameraAsync({
      allowsEditing: true,
      aspect: [4, 3],
      mediaTypes: "Images"
    });
    return result;
  }

  /**
   * Creates post object to save in itemPostings schema
   */
  createPostObject(cognitoUserId) {
    const post = {"body": this.state};
    post.body.userId = cognitoUserId;
    post.body.timeAdded = new Date().toString();
    return post;
  }

  /**
   * Saves image(s) associated with a user post to S3 bucket
   */
  saveImageToS3(cognitoUserId, imageUriArray, timeAdded) {
    imageUriArray.forEach((imageUri) => {
      Storage.put(imageUri, { 
        level: "protected",
        identityId: cognitoUserId
      })
      .then()
      .catch(error => console.log(error));
    })
  }

  /**
   * Save a new item post to postedItems schema.
   * Primary key: userId and timeAdded
   */
  async saveItemPost() {
    const userInfo = await Auth.currentUserInfo().catch(error => {
      Alert.alert(JSON.stringify(error));
      return;
    });
    const apiName = "itemPostingsCRUD";
    const path = "/itemPostings";
    const post = this.createPostObject(userInfo.id);
    API.post(apiName, path, post)
      .then(res => this.props.navigation.navigate("MainExplore", {
        reload: true
      }))
      .catch(err => console.log(err));
  }

  render() {
    return (
      <View style={{ flex: 1 }}>
        <ScrollView
          contentContainerStyle={styles.container}
          centerContent={true}
        >
          <TextInput
            onChangeText={value => this.onChangeText("itemName", value)}
            style={styles.input}
            placeholder="item name"
            placeholderTextColor="gray"
            autoCapitalize="none"
            autoCorrect={false}
          />
          <TextInput
            onChangeText={value => this.onChangeText("price", value)}
            style={styles.input}
            placeholder="price (USD)"
            placeholderTextColor="gray"
            autoCapitalize="none"
            autoCorrect={false}
          />
          <Picker
            selectedValue={this.state.category}
            style={{ height: 50, width: 100, flex: 1 }}
            onValueChange={(itemValue, itemIndex) => this.setState({ category: itemValue })}>
            <Picker.Item label="Furniture" value="furniture" />
            <Picker.Item label="Textbook" value="textbook" />
            <Picker.Item label="Electronic" value="electronic" />
          </Picker>
          <TextInput
            onChangeText={value => this.onChangeText("description", value)}
            style={styles.input}
            placeholder="description"
            placeholderTextColor="gray"
            autoCapitalize="none"
            autoCorrect={true}
          />
          <TextInput
            onChangeText={value => this.onChangeText("seller", value)}
            style={styles.input}
            placeholder="seller name"
            placeholderTextColor="gray"
            autoCapitalize="words"
            autoCorrect={false}
          />
          <Button
            style={{
              flex: 2,
              justifyContent: "center"
            }}
            color="black"
            onPress={() => this.saveItemPost()}
            title="Add Item!"
          />
        </ScrollView>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  input: {
    height: 50,
    color: "black",
    borderBottomWidth: 2,
    borderBottomColor: "teal",
    margin: 10,
    width: 300
  },
  box: {
    height: 100,
    width: 300,
    borderColor: "gray",
    borderWidth: 3
  },
  container: {
    flex: 1,
    padding: 30,
    backgroundColor: "white",
    alignItems: "center"
  },
  titleText: {
    fontSize: 36,
    padding: 50,
    color: "teal"
  }
});
