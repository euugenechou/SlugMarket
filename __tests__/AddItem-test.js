// __tests__/AddItem-test.js
import 'react-native';
import React from 'react';
import AddItem from '../src/screens/AddItem';

import renderer from 'react-test-renderer';

it('renders correctly', () => {
  const tree = renderer.create(
    <AddItem />
    ).toJSON();
  expect(tree).toMatchSnapshot();
});