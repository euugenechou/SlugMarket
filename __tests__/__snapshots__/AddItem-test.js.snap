// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  accessible={false}
  enabled={true}
  keyboardVerticalOffset={0}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "backgroundColor": "white",
        "flex": 1,
        "paddingTop": 30,
      },
      Object {
        "paddingBottom": 0,
      },
    ]
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "color": "black",
        "fontSize": 20,
        "fontWeight": "700",
        "paddingLeft": 32,
        "paddingTop": 10,
        "textAlign": "left",
      }
    }
  >
     Item Name 
  </Text>
  <TextInput
    allowFontScaling={true}
    autoCapitalize="none"
    autoCorrect={false}
    onChangeText={[Function]}
    style={
      Object {
        "alignSelf": "center",
        "borderBottomColor": "darkgray",
        "borderBottomWidth": 0.7,
        "color": "black",
        "height": 40,
        "justifyContent": "center",
        "marginBottom": 20,
        "width": 300,
      }
    }
  />
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "color": "black",
        "fontSize": 20,
        "fontWeight": "700",
        "paddingLeft": 32,
        "paddingTop": 10,
        "textAlign": "left",
      }
    }
  >
     Item Price (USD) 
  </Text>
  <TextInput
    allowFontScaling={true}
    autoCapitalize="none"
    autoCorrect={false}
    keyboardType="number-pad"
    onChangeText={[Function]}
    style={
      Object {
        "alignSelf": "center",
        "borderBottomColor": "darkgray",
        "borderBottomWidth": 0.7,
        "color": "black",
        "height": 40,
        "justifyContent": "center",
        "marginBottom": 20,
        "width": 300,
      }
    }
  />
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "color": "black",
        "fontSize": 20,
        "fontWeight": "700",
        "paddingLeft": 32,
        "paddingVertical": 10,
        "textAlign": "left",
      }
    }
  >
     Item Category 
  </Text>
  <View
    style={
      Object {
        "alignSelf": "center",
        "width": 120,
      }
    }
  >
    <RCTPicker
      items={
        Array [
          Object {
            "label": "Furniture",
            "textColor": undefined,
            "value": "Furniture",
          },
          Object {
            "label": "Textbook",
            "textColor": undefined,
            "value": "Textbooks",
          },
          Object {
            "label": "Electronic",
            "textColor": undefined,
            "value": "Electronics",
          },
        ]
      }
      onChange={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      selectedIndex={0}
      style={
        Array [
          Object {
            "height": 216,
          },
          Object {
            "fontSize": 18,
            "height": 110,
          },
        ]
      }
    />
  </View>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "color": "black",
        "fontSize": 20,
        "fontWeight": "700",
        "paddingLeft": 32,
        "paddingTop": 10,
        "textAlign": "left",
      }
    }
  >
     Item Description 
  </Text>
  <TextInput
    allowFontScaling={true}
    autoCapitalize="none"
    autoCorrect={true}
    onChangeText={[Function]}
    style={
      Object {
        "alignSelf": "center",
        "borderBottomColor": "darkgray",
        "borderBottomWidth": 0.7,
        "color": "black",
        "height": 40,
        "justifyContent": "center",
        "marginBottom": 20,
        "width": 300,
      }
    }
  />
  <View
    style={
      Array [
        Object {
          "marginLeft": 15,
          "marginRight": 15,
        },
        Object {
          "shadowColor": "rgba(0,0,0, .4)",
          "shadowOffset": Object {
            "height": 1,
            "width": 1,
          },
          "shadowOpacity": 1,
          "shadowRadius": 1,
        },
        Object {
          "alignSelf": "center",
          "paddingTop": 10,
          "width": 300,
        },
        Object {
          "borderRadius": 5,
        },
      ]
    }
  >
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={null}
    >
      <View
        pointerEvents="box-only"
        style={
          Array [
            Object {
              "alignItems": "center",
              "backgroundColor": "#9E9E9E",
              "flexDirection": "row",
              "justifyContent": "center",
              "padding": 19,
            },
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            Object {
              "backgroundColor": "teal",
            },
            Object {
              "borderRadius": 5,
            },
            Object {
              "padding": 12,
            },
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
          ]
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Array [
              Object {},
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              Array [
                Object {
                  "color": "white",
                  "fontSize": 20,
                },
                Object {
                  "color": "white",
                },
                Object {
                  "fontSize": 17.5,
                },
                undefined,
                undefined,
                Object {
                  "fontWeight": "bold",
                },
                undefined,
                undefined,
              ],
            ]
          }
        >
          Add Item
        </Text>
      </View>
    </View>
  </View>
</View>
`;
